ytrafo pipeops

sumny opened the pull request:
See #322 and #364 
This is still WIP.
Adds abstract `PipeOpInvertiblePreproc` from which you can inherit your own trafo `PipeOp`. Toy example is given in `PipeOpSimpleTrafo`. Together with `PipeOpInverter` this allows to build a `Graph` with a `Learner` in the middle, so that the target is transformed during training and retransformed for prediction. Docu is drafted; tests missing. Let's discuss this first

mb706 commented on R/PipeOpTrafo.R:
```diff
@@ -0,0 +1,240 @@
+      new_target = log(task$data(cols = task$target_names))
+      names(new_target) = paste0("log.", task$target_names)
+      task$cbind(new_target)
+      private$update_target(task, new_target = names(new_target), new_type = NULL)
```
how about a signature
```r
private$update_target(task, new_target, new_target_name = NULL, new_type = NULL)
```
and calling it here as
```r
new_target = log(task$data(cols = task$target_names))
private$update_target(task, new_target = new_target,
  new_target_name = paste0("log.", task$target_names))
```

pfistfl commented:
Please test that you can construct an arbitrary `log(base = X)` -> invert = `x ^ base` trafo.

pfistfl commented on R/PipeOpTrafo.R:
```diff
@@ -0,0 +1,240 @@
+  private = list(
+    update_target = function(task, new_target, new_type = NULL, ...) {
+      type = if (is.null(new_type)) task$task_type else assert_subset(new_type, mlr_reflections$task_types$type, empty.ok = FALSE)
+      get(mlr_reflections$task_types[type]$task)$new(id = task$id, backend = task$backend, target = new_target, ...)
```
See `convert_target` from hurdle models branch;  do not convert task if task_type stays the same.

pfistfl commented on R/PipeOpTrafo.R:
```diff
@@ -0,0 +1,240 @@
+    },
+
+    set_target_to_NA_and_possibly_convert_task_type = function(task) {
+      #FIXME: convert the task target to NA and convert the task type?
```
I guess this should just set the  `col_roles` of the original target to 0.

pfistfl commented on R/PipeOpTrafo.R:
```diff
@@ -0,0 +1,240 @@
+
+mlr_pipeops$add("inverter", PipeOpInverter)
+
+#' @title PipeOpLogtrafo
```
PipeOpSimpleTrafo:
- following params in its `param_set`: `trafo`, `inverter`, `new_target_name`, `new_task_type`
- `trafo` and `inverter` default to `identity`
- `new_target_name` defaults to `paste0("trafo_", xxold_target_name)`
- `new_task_type` defaults to `NULL` (keep old task type)

pfistfl replied:
then `PipeOpLogTrafo` could inherit from this and just define different `trafo` and `inverter`.

sumny replied:
done, maybe we discuss whether another inheritance from `PipeOpSimpleTrafo` is needed for `PipeOpLogTrafo` (it is pretty easy to set this up already in `PipeOpSimpleTrafo`)

sumny commented:
> Please test that you can construct an arbitrary `log(base = X)` -> invert = `x ^ base` trafo.

Added an example in `PipeOpSimpleTrafo`.

pfistfl commented:
Could you also add the `ppl` shortcut and also document this?

pfistfl commented on R/PipeOpNewTarget.R:
```diff
@@ -0,0 +1,43 @@
+#' Convert a task from its type to another.
+#'
+#' The task's target is replaced by a different column from the data.
+#' #FIXME: This needs detailed tests.
```
+ merge relevant tests from the old PR.

pfistfl commented on R/PipeOpNewTarget.R:
```diff
@@ -0,0 +1,43 @@
+#'   If `NULL` (default), a new task with the same task_type is created.
+#' @param drop_original_target `logical(1)`\cr
+#'   If `FALSE` (default), the original target is added as a feature.
+#'   Otherwise the original target is assigned no col_role, i.e. dropped.
```
We assign `character(0)` in order to drop col_roles

pfistfl commented on R/PipeOpTrafo.R:
```diff
@@ -0,0 +1,348 @@
+#'g$add_edge(src_id = "logtrafo", dst_id = "inverter", src_channel = 1, dst_channel = 1)
+#'g$add_edge(src_id = "logtrafo", dst_id = "regr.rpart", src_channel = 2, dst_channel = 1)
+#'g$add_edge(src_id = "regr.rpart", dst_id = "inverter", src_channel = 1, dst_channel = 2)
+#'
```
Also show the `ppl("targettrafo", ...)` version here.

sumny commented:
Renamed the `PipeOp`s, added tests, and also a `ppl` that allows for an easier build around a `PipeOpLearner` or `Graph` (where the last `PipeOp` is a `PipeOpLearner`). Also I am more strict regarding the `Task` and `Prediction` within `PipeOpTargetTrafoSimple` - I think this is currently only useful for regression tasks where the target is a (continous) numerical variable (but you can prove me wrong). This is now ready for a final review!

pfistfl commented on R/PipeOpTrafo.R:
```diff
@@ -0,0 +1,365 @@
+    },
+
+    train_target = function(task) {
+      # return a modified task and set the state
```
do we still need all the comments?

pfistfl commented on R/PipeOpTrafo.R:
```diff
@@ -0,0 +1,365 @@
+#' Only methods inherited from [`PipeOpTargetTrafo`]/[`PipeOp`].
+#'
+#' @examples
+#'library(mlr3)
```
indent here

pfistfl commented on R/mlr_graphs_elements.R:
```diff
@@ -243,3 +244,67 @@ pipeline_branch = function(graphs, prefix_branchops = "", prefix_paths = FALSE)
+
+
+
+#' @title Transform A Numerical Target And Later Invert The Transformation
```
Transform and re-transform the target variable

pfistfl commented on R/mlr_graphs_elements.R:
```diff
@@ -243,3 +244,67 @@ pipeline_branch = function(graphs, prefix_branchops = "", prefix_paths = FALSE)
+#' @title Transform A Numerical Target And Later Invert The Transformation
+#' @name mlr_graphs_targettrafo
+#' @description
+#' Creates a [`Graph`] that transforms a numerical target of a [`TaskRegr`][mlr3::TaskRegr] during
```
I would drop the numerical, this works for any target.

pfistfl replied:
Wraps a graph

pfistfl commented on R/mlr_graphs_elements.R:
```diff
@@ -243,3 +244,67 @@ pipeline_branch = function(graphs, prefix_branchops = "", prefix_paths = FALSE)
+#'  src_channel = 2, dst_channel = 1)
+#'g$add_edge(src_id = "regr.rpart", dst_id = "targetinverter",
+#'  src_channel = 1, dst_channel = 2)
+pipeline_targettrafo = function(graph) {
```
signature should be `function(graph, trafo_pipeop = NULL, id_prefix = "")`
If  `trafo_pipeop = NULL` uses `PipeOpTargetTrafoSimple`, otherwise a custom target trafo can be provided.
Additionally prefixes the `id` of trafo and inverter with `id_prefix`. 
Otherwise, this could lead to issues as duplicate id's are not allowed.

Test that this works for custom id's aswell.

pfistfl commented on R/PipeOpNewTarget.R:
```diff
@@ -0,0 +1,43 @@
+#' Convert a task from its type to another.
```
File Name is wrong

pfistfl commented on tests/testthat/test_pipeop_newtarget.R:
```diff
@@ -0,0 +1,20 @@
+test_that("convert_task reconstructs task", {
```
could you copy all relevant tests?  (Or possibly test a few more edge cases).

In theory, this function should also be able to just change iris into a regr task by predicting `Sepal.Length` instead of `Species` etc. guess we  can do a few more sensible tests.

pfistfl commented on tests/testthat/test_pipeop_targettrafosimple.R:
```diff
@@ -0,0 +1,74 @@
+context("PipeOpTargetTrafoSimple")
```
So the following currently breaks, as it would require a `PredictionClassif` -> `PredictionRegr` conversion
- Task: `boston_housing`, `trafo = x > 20`,  inverter if (x) 19 else 21 (then we'd need a classif learner)

While the following might work. This is actually somewhat relevant.
- Task: `boston_housing`,`trafo = x > 20`,  inverter `identity`? This should probably be possible.

pfistfl replied:
Could we test that and possibly add a fixme that we might somehow enable the first option in the future?

sumny replied:
I added two more tests here, one that goes from  `PredictionClassif` -> `PredictionRegr` and one the other way. Both are not the most meaningful ones but it is possible. For most cases to actually be meaningful the user will rely on setting the `$state` and rely on this for the inversion (like statet by you below).

pfistfl commented on tests/testthat/test_pipeop_targettrafosimple.R:
```diff
@@ -0,0 +1,74 @@
+ learner_predict_out = learner$predict(task_log)
+ expect_equal(2 ^ learner_predict_out$truth, predict_out[[1L]]$truth)
+ expect_equal(2 ^ learner_predict_out$response, predict_out[[1L]]$response)
+})
```
We currently have no checks for the state, i.e. whether learning and applying a state possibly works.

I guess we could do one `PipeOp` that covers this in the PR's comments.

sumny replied:
yes, seems like a good idea!

pfistfl commented:
We currently can not test that the `state` actually works as we have no `PipeOp` that uses it.

I will describe a `PipeOp` we should implement here in order to showcase how such pipeops are implemented and to test that this works in general.
```
PipeOpTargetTrafoScaleRange
...
basically does the same as `PipeOpScaleRange`, i.e. stores range trafo to state.
additionally, it knows how to retransform scaled data.
```

sumny commented:
> We currently can not test that the `state` actually works as we have no `PipeOp` that uses it.
> 
> I will describe a `PipeOp` we should implement here in order to showcase how such pipeops are implemented and to test that this works in general.
> 
> ```
> PipeOpTargetTrafoScaleRange
> ...
> basically does the same as `PipeOpScaleRange`, i.e. stores range trafo to state.
> additionally, it knows how to retransform scaled data.
> ```

I added a `PipeOpTargetTrafoScaleRange` and tests. Seems to work nice. Had to add a `set_state()` method to `PipeOpTargetTrafo` that can be overloaded that makes such trafos way easier. `set_state()` is called a single time in `.train()`, before `train_task()` is evaluated so you can then rely on the `$state` both during training and prediction.
